### PR TITLE
Kinesis Consumer Library for Kinesis streaming

### PR DESCRIPTION
**Pull request by kiambogo**

### CODE REVIEW

**Comment by dmateusp** [comment]
#18

File: src/main/scala/fs2/aws/kinesis/CommittableRecord.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package fs2
+package aws
+package kinesis
+
+import com.amazonaws.services.kinesis.clientlibrary.types.{ ExtendedSequenceNumber, UserRecord }
+import com.amazonaws.services.kinesis.model.Record
+import com.amazonaws.services.kinesis.clientlibrary.interfaces.IRecordProcessorCheckpointer
+
+
+case class CommittableRecord(
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/CommittableRecord.scala]: scala docs here please :)

File: src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package fs2
+package aws
+package kinesis
+
+import scala.concurrent.duration._
+
+case class KinesisWorkerStreamSettings(
+  bufferSize: Int,
+  backpressureTimeout: FiniteDuration) {
+  require(
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala]: is that good practice ? Should you not have a class and a companion object with a constructor that returns an Either instead?

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import com.amazonaws.services.kinesis.clientlibrary.types._
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.ShutdownReason
+import scala.collection.JavaConverters._
+
+
+private[kinesis] class RecordProcessor(cb: CommittableRecord => Unit) extends v2.IRecordProcessor {
+   private var shardId: String = _
+   private var extendedSequenceNumber: ExtendedSequenceNumber = _
+   var shutdown: Option[ShutdownReason] = None
+   var latestCheckpointer: Option[IRecordProcessorCheckpointer] = None
```
Comment (mmienko) [inline src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: I think there's different spacing here, at least seems so on Github.

Comment (mmienko) [reply src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: or it's the def's underneath,  not sure which is right.

File: src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package fs2
+package aws
+package kinesis
+
+import scala.concurrent.duration._
+
+case class KinesisWorkerStreamSettings(
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala]: Rename the file "KinesisSettings" ? there are all your Settings models here not only "KinesisWorkerSettings"

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+package kinesis
+
+import com.amazonaws.services.kinesis.clientlibrary.interfaces._
+import com.amazonaws.services.kinesis.clientlibrary.types._
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.ShutdownReason
+import scala.collection.JavaConverters._
+
+
+private[kinesis] class RecordProcessor(cb: CommittableRecord => Unit) extends v2.IRecordProcessor {
+   private var shardId: String = _
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: do we need all those to be vars?

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+package aws
+package kinesis
+
+import com.amazonaws.services.kinesis.clientlibrary.interfaces._
+import com.amazonaws.services.kinesis.clientlibrary.types._
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.ShutdownReason
+import scala.collection.JavaConverters._
+
+
+private[kinesis] class RecordProcessor(cb: CommittableRecord => Unit) extends v2.IRecordProcessor {
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: some scaladocs ? :)

File: src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package fs2
+package aws
+package kinesis
+
+import scala.concurrent.duration._
+
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala]: some docs about the package ?

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+   private var shardId: String = _
+   private var extendedSequenceNumber: ExtendedSequenceNumber = _
+   var shutdown: Option[ShutdownReason] = None
+   var latestCheckpointer: Option[IRecordProcessorCheckpointer] = None
+
+  def getShardId = shardId
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: do we need getters?

File: src/main/scala/fs2/aws/kinesis/kcl.scala [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import com.amazonaws.services.kinesis.model.Record
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker
+import com.amazonaws.services.kinesis.clientlibrary.interfaces.v2.IRecordProcessorFactory
+
+package object kcl {
+  /** Intialize a worker and start streaming records from a Kinesis stream
+    * On stream finish (due to error or other), worker will be shutdown
+    *
+    *  @tparam F effect type of the fs2 stream
+    *  @param workerFactory function to create a Worker from a IRecordProcessorFactory
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/kcl.scala]:
```suggestion
    *  @param workerFactory function to create a [[com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker]] from a [[com.amazonaws.services.kinesis.clientlibrary.interfaces.v2.IRecordProcessorFactory]]
```

File: src/main/scala/fs2/aws/kinesis/kcl.scala [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    *  @param config configuration settings for the KCL worker
+    *  @return an infinite fs2 Stream that emits Kinesis Records
+    */
+  def readFromKinesisStream[F[_]](
+    workerFactory: => IRecordProcessorFactory => Worker,
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/kcl.scala]: I'm not familiar with Kinesis enough but is there a way we could provide a default workerFactory ? or abstract completely the need for a factory in the first place ? Factory sounds very Javaish

File: src/main/scala/fs2/aws/kinesis/kcl.scala [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+      .map { case (k, st) =>
+        st.tupleLeft(k)
+          .groupWithin(checkpointSettings.maxBatchSize, checkpointSettings.maxBatchWait)
+          .map(_.toList.max)
+          .mapAsync(1) { cr =>
+            F.async[Record](_ =>
+              if (cr._2.canCheckpoint)
+                Right(cr._2.checkpoint)
+              else
+                Left(throw new Exception("Record processor has been shutdown - cannot checkpoint records"))
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/kcl.scala]: Can we create an exceptions package in internal and return "narrower" Exceptions instead?

File: src/main/scala/fs2/package.scala [hunk 12]
```diff
@@ -0,0 +1,39 @@
+package fs2
```
Comment (dmateusp) [inline src/main/scala/fs2/package.scala]: weird, it's a new package, but we have an "internal" package already under src/main/scala/fs2/aws/internal

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala (hunk 3 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: fixed

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala (hunk 5 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: this class extends the AWS record processor java abstract class. the `initialize` method receives the input which dictates which shard this record processor is for. do you have an idea on how to capture the data we need and avoid using vars? It would be great to keep it immutable

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala (hunk 8 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: nope :)

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala (hunk 6 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: :you_got_it_dude:

File: src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala (hunk 2 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/kinesis/KinesisWorkerSettings.scala]: good question! I'm not sure on what the best practice is for scala around the use of `require` and `assert`. Either way, I think its more functional to return Either, so I'll go down that route

File: src/main/scala/fs2/aws/kinesis/RecordProcessor.scala (hunk 5 above)
Comment (dmateusp) [reply src/main/scala/fs2/aws/kinesis/RecordProcessor.scala]: I'll have to pull it down and play with it!

File: src/test/scala/KinesisConsumerSpec.scala [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+          rp.processRecords(recordsInput.withRecords(List(record).asJava))
+        }
+      }
+    }
+
+    simulateWorkerThread(recordProcessor)
+    simulateWorkerThread(recordProcessor2)
+
+    // Should process all 50 messages
+    eventually(output.size shouldBe(60))
```
Comment (semenodm) [inline src/test/scala/KinesisConsumerSpec.scala]: 50 in the comment

File: src/main/scala/fs2/aws/kinesis/kcl.scala (hunk 10 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/kinesis/kcl.scala]: Good idea! I'll play around with something a bit more idiomatic

File: src/test/scala/KinesisConsumerSpec.scala (hunk 13 above)
Comment (kiambogo) [reply src/test/scala/KinesisConsumerSpec.scala]: Right - it does 10 at first, then an additional 50, so 60 total. Comment could be made better

File: src/main/scala/fs2/aws/internal/Exceptions.scala [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package fs2
+package aws
+package internal
+
+object Exceptions {
+  case class KinesisCheckpointException(msg: String) extends Exception
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/internal/Exceptions.scala]: nice!

File: src/main/scala/fs2/aws/kinesis/CommittableRecord.scala (hunk 1 above)
Comment (dmateusp) [reply src/main/scala/fs2/aws/kinesis/CommittableRecord.scala]: nice!

File: src/main/scala/fs2/aws/kinesis/KinesisSettings.scala [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    val maxBatchSize: Int,
+    val maxBatchWait: FiniteDuration
+)
+
+object KinesisStreamSettings {
+  val defaultInstance: KinesisStreamSettings = new KinesisStreamSettings(10)
+
+  def apply(bufferSize: Int): Either[Throwable, KinesisStreamSettings] =
+    if (bufferSize < 1)
+      Left(new IllegalArgumentException("Buffer size must be greater than 0"))
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/KinesisSettings.scala]: Could be a KinesisSettingsException !

File: src/main/scala/fs2/aws/kinesis/KinesisSettings.scala [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    else
+      Right(new KinesisStreamSettings(bufferSize))
+}
+
+object KinesisCheckpointSettings {
+  val defaultInstance = new KinesisCheckpointSettings(1000, 10.seconds)
+
+  def apply(maxBatchSize: Int,
+            maxBatchWait: FiniteDuration): Either[Throwable, KinesisCheckpointSettings] =
```
Comment (dmateusp) [inline src/main/scala/fs2/aws/kinesis/KinesisSettings.scala]: There are multiple types of KinesisSettings Exception so I think you can create a sealed trait KinesisSettingsException and then case object / case classes that represent the specific errors

Comment (kiambogo) [reply src/main/scala/fs2/aws/kinesis/KinesisSettings.scala]: neat, sounds good

File: src/main/scala/fs2/aws/internal/Internal.scala [hunk 17]
```diff
@@ -46,4 +49,38 @@ object Internal {
...
+
+      (in ++ Stream.eval_(cleanup))
+        .evalMap { el =>
+          (selector(el), st.get).mapN { (key, queues) =>
+            queues
+              .get(key)
+              .fold {
+                for {
+                  newQ <- Queue.unbounded[F, Option[A]] // Create a new queue
+                  _    <- st.modify(x => (x + (key -> newQ), x)) // Update the ref of queues
```
Comment (semenodm) [inline src/main/scala/fs2/aws/internal/Internal.scala]: Can we rename this magic `x` to `shardsToQueues` or something meaningful, it is hard to follow

File: src/main/scala/fs2/aws/internal/Internal.scala [hunk 18]
```diff
@@ -46,4 +49,38 @@ object Internal {
...
+                  _    <- newQ.enqueue1(el.some)
```
Comment (semenodm) [inline src/main/scala/fs2/aws/internal/Internal.scala]: `e1.some` is it lifting `A` to `Option[A]` ?

File: src/main/scala/fs2/aws/internal/Internal.scala (hunk 17 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/internal/Internal.scala]: yes, i'll try to make it a bit more clear

File: src/main/scala/fs2/aws/internal/Internal.scala (hunk 18 above)
Comment (kiambogo) [reply src/main/scala/fs2/aws/internal/Internal.scala]: yep - exposed by cats implicits

**Comment by semenodm** [comment]
LGTM
